### PR TITLE
Flash Map: add shell command to list flash areas

### PR DESCRIPTION
**Pull request by nashif**
Minor modifications to flash map API adding doxygen comments and add a new shell to list flash areas if the device has those.

examples:

```
uart:~$ flash_map list
ID | Device | Device Name       |   Offset   |   Size
-------------------------------------------------------
0    0        NRF_FLASH_DRV_NAME    0            49152
1    0        NRF_FLASH_DRV_NAME    479232       430080
2    0        NRF_FLASH_DRV_NAME    49152        430080
3    0        NRF_FLASH_DRV_NAME    909312       122880
4    0        NRF_FLASH_DRV_NAME    1032192      16384
```

### CODE REVIEW

File: include/flash_map.h [hunk 1]
```diff
@@ -50,34 +55,98 @@ struct flash_sector {
...
 
-/*
+/**
  * Given flash area ID, return info about sectors within the area.
+ *
+ * @param fa_id Flash area ID
+ * @param count Sector count
+ * @param sectors Flash sectors
+ *
+ * @return 1 On success. -ENODEV if no driver match.
```
Comment (nvlsianpu) [inline include/flash_map.h]:
```suggestion
@return 0 on success. -ENOMEM if There are too many flash pages on the device to fit in the array. -EINVAL if no flash area match.
```

File: include/flash_map.h [hunk 2]
```diff
@@ -50,34 +55,98 @@ struct flash_sector {
...
+ * @return Block size
  */
 u8_t flash_area_align(const struct flash_area *fa);
 
-/*
+/**
  * Given flash area ID, return info about sectors within the area.
+ *
+ * @param fa_id Flash area ID
+ * @param count Sector count
```
Comment (nvlsianpu) [inline include/flash_map.h]: Capacity of sectors array on input. Exists sectors count on successful output.

**Review by nvlsianpu — DISMISSED** [review]
Shell command code looks good. I have comments only to documentation.

File: include/flash_map.h (hunk 2 above)
Comment (nvlsianpu) [reply include/flash_map.h]: second sentence is stil valid.

**Comment by nvlsianpu** [comment]
Think you need to squash review changes,

**Comment by nashif** [comment]
> Think you need to squash review changes,

they are squashed already

File: subsys/storage/flash_map/Kconfig [hunk 3]
```diff
@@ -15,10 +15,18 @@ menuconfig FLASH_MAP
 	help
 	  Enable support of flash map abstraction.
 
+if FLASH_MAP
+
+config FLASH_MAP_SHELL
+	bool "Enable flash map shell interface"
+	help
```
Comment (aurel32) [inline subsys/storage/flash_map/Kconfig]: This is missing a "`depends on SHELL`" line.

File: include/flash_map.h (hunk 2 above)
Comment (nashif) [reply include/flash_map.h]: I am not sure exactly what you mean.

**Review by nvlsianpu — APPROVED** [review]

File: subsys/storage/flash_map/flash_map_shell.c [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+
+LOG_MODULE_REGISTER(flash_map_shell);
+
+extern const struct flash_area *flash_map;
+
+static void fa_cb(const struct flash_area *fa, void *user_data)
+{
+	struct shell *shell = user_data;
+
+	shell_print(shell, "%-4d %-8d %-20s  %-12u %-12zu",
```
Comment (nvlsianpu) [inline subsys/storage/flash_map/flash_map_shell.c]: just nip, not a issue - I think hex values are more useful/readable than decimal. Most of memories has power of 2 internal subdivision. So maybe worth to change this.

**Review by nvlsianpu — COMMENTED** [review]
Just checked how it works on nrf52840 - fine. Thanks for this patch.

File: include/flash_map.h [hunk 5]
```diff
@@ -33,8 +33,13 @@ extern "C" {
 #include <stddef.h>
 #include <sys/types.h>
 
-#define SOC_FLASH_0_ID 0	/* device_id for SoC flash memory driver */
-#define SPI_FLASH_0_ID 1	/* device_id for external SPI flash driver */
+/**
+ * @brief Flash Area
+ * @defgroup flashmap_api Flah Area
```
Comment (galak) [inline include/flash_map.h]: Fix typo, `Flah' -> `Flash`.  Also do you mind marking this as deprecated.

Comment (nashif) [reply include/flash_map.h]: what should be deprecated?

File: include/flash_map.h (hunk 2 above)
Comment (nvlsianpu) [reply include/flash_map.h]: I omitted your response. I mean that count is [in, out] parameter:
```suggestion
 * @param[in, out] count Capacity of `sectors` array on input. Count of fetched sectors on output.
 * @param[out] sectors array
```

**Review by galak — APPROVED** [review]

**Pull request merged**